src/compat/db_json.cpp

```diff
@@ -2571,3 +2733,504 @@ JSON_WALKER::WalkValue (JSON_VALUE &value)
+  str_length = or_get_int (buf, &rc);
+  if (rc != NO_ERROR)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_TF_BUFFER_OVERFLOW, 0);
+      return rc;
+    }
+
+  rc = db_json_or_buf_underflow (buf, str_length);
+  if (rc != NO_ERROR)
+    {
+      // we need to assert error here because or_underflow sets the error unlike or_overflow
+      // which only returns the error code
+      ASSERT_ERROR();
+      return rc;
+    }
+
+  // set the string directly from the buffer to avoid additional copy
+  value.SetString (buf->ptr, str_length - 1, doc_allocator);
+  // update the buffer pointer
+  buf->ptr += str_length;
+
+  // still need to take care of the alignment
+  rc = or_align (buf, INT_ALIGNMENT);
+  if (rc != NO_ERROR)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_TF_BUFFER_OVERFLOW, 0);
+      return rc;
+    }
+
+  return NO_ERROR;
+}
+
+static int
+db_json_unpack_int_to_value (OR_BUF *buf, JSON_VALUE &value)
+{
+  int rc = NO_ERROR;
+  int int_value;
+
+  // unpack int
+  int_value = or_get_int (buf, &rc);
+  if (rc != NO_ERROR)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_TF_BUFFER_OVERFLOW, 0);
+      return rc;
+    }
+
+  value.SetInt (int_value);
+
+  return NO_ERROR;
+}
+
+static int
+db_json_unpack_double_to_value (OR_BUF *buf, JSON_VALUE &value)
+{
+  int rc = NO_ERROR;
+  double double_value;
+
+  // unpack double
+  double_value = or_get_double (buf, &rc);
+  if (rc != NO_ERROR)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_TF_BUFFER_OVERFLOW, 0);
+      return rc;
+    }
+
+  value.SetDouble (double_value);
+
+  return NO_ERROR;
+}
+
+static int
+db_json_unpack_bool_to_value (OR_BUF *buf, JSON_VALUE &value)
+{
+  int rc = NO_ERROR;
+  int int_value;
+
+  int_value = or_get_int (buf, &rc); // it can be 0 or 1
+  if (rc != NO_ERROR)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_TF_BUFFER_OVERFLOW, 0);
+      return rc;
+    }
+
+  value.SetBool (int_value == 1 ? true : false);
+
+  return NO_ERROR;
+}
+
+static int
+db_json_unpack_object_to_value (OR_BUF *buf, JSON_VALUE &value, JSON_PRIVATE_MEMPOOL &doc_allocator)
+{
+  int rc = NO_ERROR;
+  int size;
+
+  value.SetObject();
+
+  // get the member count of the object
+  size = or_get_int (buf, &rc);
+  if (rc != NO_ERROR)
+    {
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_TF_BUFFER_OVERFLOW, 0);
+      return rc;
+    }
+
+  // for each key-value pair we need to deserialize the value
+  for (int i = 0; i < size; i++)
+    {
+      // get the key
+      JSON_VALUE key;
+      rc = db_json_unpack_string_to_value (buf, key, doc_allocator);
+      if (rc != NO_ERROR)
+	{
+	  ASSERT_ERROR();
+	  return rc;
+	}
+
+      // get the value
+      JSON_VALUE child;
+      rc = db_json_deserialize_doc_internal (buf, child, doc_allocator);
+      if (rc != NO_ERROR)
+	{
+	  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_TF_BUFFER_OVERFLOW, 0);
+	  return rc;
+	}
+
+      value.AddMember (key, child, doc_allocator);
```

Comment:
Does value.AddMember copy key and child again? I think not, but can you please confirm?

Reply:
AddMember and PushBack do not make additional copies but only assign raw pointer to the data. The copy is made, for example, explicitly in SeString.